### ISSUE TITLE
OWL benchmark

### ISSUE THREAD

**Issue by woodholly** [odoo/owl#1539]
Hi guys!
I have added a benchmark for OWL framework.
Benchmark is based on well known js-framework-benchmark. My branch with OWL added:
https://github.com/woodholly/js-framework-benchmark/tree/owl/frameworks/keyed/owl/src

Results:
Owl is almost as fast as Vue but has a flaw in 1 test. "select row" test is 2 times slower than Vue/React.
I think you could be interested.

BTW as far as I know, OWL does not support non-keyed mode (details on what is keyed/non-keyed mode are in js-framework-benchmark docs), and keys are always required, right?

**Comment by aadrian** [comment]
@woodholly I guess the main issue with the benchmark, to be really useful: it might need the random data to come from a server (not generated in the browser).

**Comment by ged-odoo** [comment]
@aadrian it's true that using data coming from a server would be more relevant, as a realistic usecase. but the js-framework-benchmark is basically a benchmark on rendering speed, so I think that it is still interesting

@woodholly I am slightly disappointed and proud at the same time of the results you got. Slightly disappointed because I think that Owl is world class, and the underlying engine is faster than Vue (but Vue implements v-memo, more on that below). Also, I actually made the same benchmark myself 6 months ago, and I got a geometric average of 1.15, so, faster than Vue. Maybe Owl got slightly slower because we added a few features and fixes. Also, another thing that you can do:

1. use `.synthetic` on your event handlers:
```xml
                    <a t-on-click.synthetic="() => this.selectRow(row.id)"><t t-esc="row.label"/></a>
```

2. use raw state instead of reactive state (so, without a `useState`). Then, you need to call manually `this.render()` whenever you update the state. By doing this, you can compare the cost of reading your state through a reactive object.

Doing these 2 things should bring a little bit more speed. If you do it, I will be interested to know if it makes a difference. I have a feeling that the reactivity code is quite hot, and can be (slightly) optimized.

But now, most vdom-based frameworks are slow on the select row benchmark. But vue is really fast, and the real reason for that is because it implements a `v-memo` directive, which allow Vue to bypass generating and updating a sub part of the DOM. So, having similar results as Vue in Owl, but without `v-memo` is already a big achievement. Note that I actually implemented an equivalent `t-memo` directive in Owl, and it does bring a significant speed improvement on that benchmark.  However, I decided against adding it to Owl for a few reasons:

- vdom-based frameworks have already a standard way of memoizing sub parts of the interface: components. It is telling that almost all framework benchmark implementations use a single component: it is faster for the benchmark. But in a real application, you would use sub components.
- v-memo is kind of unsafe: it is very easy to add implicit dependencies and forget to update the dependency array. It will look like it works, but then weird bugs will happen. It is particularly relevant in an environment like Odoo, in which we have thousands of developers working on various part of the code that need to keep working as expected.
- if one really encounters such a performance problem, then one can still perform manual optimizations (opting out of reactivity, using synthetic events, or even implementing the hot codepath in pure js, without using owl).

So, in summary, I am not a fan of `v-memo`. We could have it in Owl, and maybe someday, we will do it to improve our benchmark numbers, but I feel that it is not representative of a real usecase. 

Finally, yes, owl only works in keyed mode. I  think that the js-framework-benchmark should stop benchmarking unkeyed implementations. It is possible to have very fast numbers by doing dangerous stuff that should never happen in a real framework (like reusing discarded html elements)

**Comment by sdegueldre** [comment]
Thanks for doing tihs, it looks like there's nothing actionable on our part so I'm closing this issue but feel free to keep discussing the benchmark stuff on it.
